Add `@content` support to `@tailwindcss/cli`

RobinMalfait opened the pull request:
This is a continuation of #14048 where we specifically focus on the `@tailwindcss/cli`.

This PR optimizes the globs such that we always have a clean `base` and the `glob` (which is the first dynamic part).

This PR also refactors the CLI such that we can watch files defined by the `@content` at-rules in the CSS. The `@parcel/watcher` watches from a base path (not a glob), which means that we have to create a watcher for each glob's base.

This PR also adds a `scanFiles` to the `scanDir` result, so that we can verify that a scanned file matches the globs we already know from `scanDir`.

<!--

👋 Hey, thanks for your interest in contributing to Tailwind!

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a lot of time and effort into a new feature. To avoid this from happening, we request that contributors create an issue to first discuss any significant new features. This includes things like adding new utilities, creating new at-rules, or adding new component examples to the documentation.

https://github.com/tailwindcss/tailwindcss/blob/master/.github/CONTRIBUTING.md

-->

RobinMalfait committed:
add disposables

RobinMalfait committed:
remove destructuring

This will use a variable to hold everything instead of destructuring
such that it will be easier to completely replace them with new
instances.

RobinMalfait committed:
ensure we are working with the new input and compiler

Otherwise `compiler.globs` is referring to the older compiler's globs
and `input` refers to the older `input` as well.

This commit now makes sure that we are working with the new input and
compiler before re-compiling the CSS.

RobinMalfait committed:
create watchers for each glob's base

The `@parcel/watcher` watches a specific base folder, so it's up to us
to compute the base of each glob and use that as the base.

This means that we will have multiple watcher instances that we need to
manage.

Right now we will make sure that:
1. All watchers are cleaned up when the process exits
2. All watchers are cleaned up when we have to perform a full rebuild
   (e.g.: a CSS file changed)
    1. This is necessary because there could be a change in `@content`,
       and we need to make sure that we remove old globs and only use
       the new globs.

RobinMalfait committed:
attach `scanFiles` directly on the scan dir result

When a new file is discovered, we need to make sure that it actually
is allowed to be analyzed. To do this, we have to make sure that the
file matches the `globs` we know about.

We could scan the globs and pull in a library to verify that the file is
allowed.

Alternatively we could create a function in Rust that receives the files
and the list of globs. While this works, this means that we have to send
the globs every time we want to scan a new file.

Instead, this attaches a `scanFiles` directly to the `scanDirResult`
which means we can access the `globs` from scope in Rust without
sending it "over the wire".

RobinMalfait committed:
bail when no new candidates are found

The watcher has to watch a base directory, if we see a changed file we
verify that it matches globs before we try to extract candidates.

If the files doesn't match any of the globs, then we can bail early.
This will result in no visual output which is probably what the user
expects.

However, if a valid file results in no candidates then we also bail
early and don't show anything. Maybe that's confusing, but nothing new
happened and the CSS won't change since it's the same.

RobinMalfait committed:
dedupe events

We currently handle `create` and `update` events the same, and we don't
care about `delete` events for now.

Therefore we can just keep track of a set of files that were created or
updated.

RobinMalfait committed:
check for cssImportPaths first

This saves us a tiny bit of work when a `.css` is changed because we
will break and do a full rebuild anyway.

RobinMalfait committed:
use a macrotask instead of microtask

If you happen to have globs that match the same base path, it could
result in a file that is picked up by multiple watchers. We used a
microtask to collect all of them but that's not enough.

Instead, at least on macOS, a macrotask is enough (implemented as a
`setTimeout(cb, 0)`). Even if we trigger a build multiple times it
should result in a no-op the second time so it's not the end of the
world.

This is a bit nicer, because now you only see a single output log in
instead of `n`.

RobinMalfait committed:
revert unnecessary change

RobinMalfait commented on crates/node/src/lib.rs:
```diff
@@ -22,20 +22,40 @@ impl From<ChangedContent> for tailwindcss_oxide::ChangedContent {
 
 #[derive(Debug, Clone)]
-#[napi(object)]
+#[napi]
```
We are attaching a new function `scan_files`, so the simpler `#[napi(object)]` is not enough anymore. Using `#[napi(object)]` results in the missing function.

RobinMalfait commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -218,15 +215,28 @@ export async function handle(args: Result<ReturnType<typeof options>>) {
+          compiler = compile(input)
+
+          // Re-scan the directory to get the new `candidates`
+          scanDirResult = scanDir({ base, contentPaths: compiler.globs, outputGlobs: true })
+
+          // Setup new watchers
+          cleanupWatchers = await createWatchers(scanDirResult.globs, handle)
```
We need to know the new globs before we can setup the new watchers. This also means that we technically have a potential race condition if a file changed between these 2 function calls.

One solution could be to keep the previous watchers alive and re-trigger a change when new files are created or existing files updated. Once the new watchers are created, we could cleanup the old ones.

Another solution could be to run `scanDir` _again_ once the watchers are setup. That way we will have all the missing candidates if a new file was added in the meantime.


Is this something we should try to figure out already? 🤔

philipp-spiess replied:
Hmm, probably? I feel like reliability in the dev setup is super important. 

> Another solution could be to run `scanDir` again once the watchers are setup.

I like this approach. It shouldn't yield files that are invalid and since we check modification date on the rust side this should no-op for the most part, right?

RobinMalfait replied:
Note: For the initial build, we made sure to setup the watchers before calling the actual `compiler.build()` step to ensure that we don't miss any changes in the meantime.

For this particular case, this re-creating all the watchers step will only ever happen if the CSS (or any of its imported CSS files) changes. I don't think it's worth to try and solve this issue until we have proof that somebody runs into this issue.

I think it will only happen if you have a CSS file, you change it (but don't save yet), and you have a new unsaved file that will match any of the globs in the CSS file. Then you use a "save all" feature in your editor. If then the save of the new file is delayed just long enough, then we could run into this issue, maybe.

(Discussing this with @philipp-spiess on a call right now. Writing this for future reference purposes)

RobinMalfait commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -218,15 +215,28 @@ export async function handle(args: Result<ReturnType<typeof options>>) {
         // Scan changed files only for incremental rebuilds.
         else if (rebuildStrategy === 'incremental') {
-          let newCandidates = scanFiles(changedFiles, IO.Sequential | Parsing.Sequential)
+          let newCandidates = scanDirResult.scanFiles(changedFiles)
```
Attached this `scanFiles` method to the `scanDirResult` such that we don't have to pass in all the globs to ensure that a file is allowed or not.

This way we can keep the state (the globs) in Rust

philipp-spiess replied:
Beautiful!

RobinMalfait commented on packages/@tailwindcss-postcss/src/index.test.ts:
```diff
@@ -245,8 +245,8 @@ describe('@content', () => {
       type: 'dir-dependency',
-      dir: expect.stringMatching(/example-project$/),
-      glob: '../other-project/src/**/*.js',
+      dir: expect.stringMatching(/other-project\/src$/),
+      glob: '**/*.js',
```
We are optimizing the globs using the `get_fast_patterns` where we try to move as many static parts to the "base" of the glob, and only keep the dynamic parts in the glob.

In this case, `../other-project/src/` is attached to the base (and later resolved such that the `..` is gone). Then the `**/*.js` stays in the glob.

philipp-spiess commented on crates/oxide/src/glob.rs:
```diff
@@ -136,6 +139,14 @@ fn get_fast_patterns(base_path: &Path, patterns: &Vec<String>) -> Vec<(PathBuf,
 }
 
+pub fn path_matches_globs(path: &Path, globs: &[GlobEntry]) -> bool {
+    let path = format!("{}", path.display());
```
The intention here is just to convert the `path` to string, right? Perhaps we can use `Path#to_string_lossy` instead to make it clearer and avoid the intermediate display representation?

philipp-spiess commented on crates/oxide/src/glob.rs:
```diff
@@ -136,6 +139,14 @@ fn get_fast_patterns(base_path: &Path, patterns: &Vec<String>) -> Vec<(PathBuf,
     optimized_patterns
 }
 
+pub fn path_matches_globs(path: &Path, globs: &[GlobEntry]) -> bool {
```
I feel like this is going to be useful for a bunch of other things as well (e.g. the `node_module` should warn unless the glob it matches contains `node_modules` idea we chatted about last week).

philipp-spiess commented on crates/oxide/src/lib.rs:
```diff
@@ -429,6 +447,26 @@ pub fn scan_files(input: Vec<ChangedContent>, options: u8) -> Vec<String> {
 }
 
+#[tracing::instrument(skip(input, globs))]
+pub fn scan_files_with_globs(input: Vec<ChangedContent>, globs: Vec<GlobEntry>) -> Vec<String> {
```
Where will we need this?

philipp-spiess replied:
Update: This helps with testability

philipp-spiess commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -255,6 +266,44 @@ export async function handle(args: Result<ReturnType<typeof options>>) {
+  function flush() {
+    d.dispose()
+    d.queueMacrotask(() => {
+      handle(Array.from(files))
```
If we have two globs yielding the same file, the idea here is that we only emit the files once, right?

I think we should check if files is non-empty before we call handle otherwise we will still have two callback invocations (even though only one contains the changed/added file)

philipp-spiess replied:
Nevermind, this is what disposables is for and why we call `dispose()` before queueing another macrotask. This will cancel all previously scheduled macro tasks.

philipp-spiess commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -255,6 +266,44 @@ export async function handle(args: Result<ReturnType<typeof options>>) {
+  let files = new Set<string>()
+
+  let d = disposables()
+  function flush() {
```
Mega nit but `flush` feels like it should do something synchronously. Maybe prefixing with "enqueue" could help. What do you think?

philipp-spiess reviewed (commented):
Awesome! This is great. I left a few comments inline.

philipp-spiess commented on crates/node/src/lib.rs:
```diff
@@ -22,20 +22,40 @@ impl From<ChangedContent> for tailwindcss_oxide::ChangedContent {
+#[napi]
+impl ScanResult {
+  #[napi]
+  pub fn scan_files(&self, input: Vec<ChangedContent>) -> Vec<String> {
```
ah nevermind :D

RobinMalfait committed:
Rename `flush` to `enqueueFlush`

`flush` is not synchronous, but the function name looks synchronous.
Let's make it feel a bit more async.

RobinMalfait committed:
setup watchers before the initial build

This way we make sure we don't miss any changes between the initial scan
and the initial build.

RobinMalfait committed:
setup CLI tests

This currently invokes the `build` command directly. In a future PR we
will add proper integration tests where the CLI is tested from the
outside.

For now, this will do.

RobinMalfait committed:
collect `globs`, and pass them to `scanDir`

thecrypticace commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -207,15 +208,28 @@ export async function handle(args: Result<ReturnType<typeof options>>) {
+          compiler = compile(input)
+
+          // Re-scan the directory to get the new `candidates`
+          scanDirResult = scanDir({ base, contentPaths: compiler.globs })
```
Should we make `contentPaths` a `Vec<GlobEntry>` or something like that and do any merging / concatenation on the Rust side? I think it'd be ideal that the base and glob are separate for as long as possible.

RobinMalfait replied:
Yep, we ended up doing exactly this, because we ran into an issue: https://github.com/tailwindlabs/tailwindcss/pull/14067#discussion_r1695607073

Essentially `base` was referring to the `process.cwd()` (for auto content detection. But globs were relative to the base path of the `input` file, which could live in `process.cwd() + './src/styles'` for example.

`Vec<GlobEntry>`, makes a lot of sense here thanks!

adamwathan commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -235,13 +249,59 @@ export async function handle(args: Result<ReturnType<typeof options>>) {
+  eprintln(`Done in ${formatDuration(end - start)}`)
+}
+
+async function createWatchers(globs: GlobEntry[], handle: (files: string[]) => void) {
+  let watchers = disposables()
+  let files = new Set<string>()
+
+  let d = disposables()
+  function enqueueFlush() {
+    d.dispose()
+    d.queueMacrotask(() => {
+      handle(Array.from(files))
+      files.clear()
+    })
+  }
+
+  // Setup a watcher for every glob based on the `base` directory.
+  for (let glob of globs) {
+    // Globs with `!` are negated and should not require a dedicated watcher.
+    if (glob.glob[0] === '!') continue
+
+    let { unsubscribe } = await watcher.subscribe(glob.base, async (err, events) => {
+      if (err) {
+        console.error(err)
+        return
+      }
+
+      for (let event of events) {
+        if (event.type === 'delete') continue
+        files.add(event.path)
+      }
+
+      enqueueFlush()
+    })
+
+    watchers.add(unsubscribe)
+  }
+
+  return () => watchers.dispose()
```
I think this function could use more documentation — there's a lot of little details and stuff in here that aren't really explained at the moment, like what `enqueueFlush` is for for example

RobinMalfait committed:
keep glob as-is

Let's not join the basePath and glob together at this time. The glob is
a relative path and it's relative to the dirname of where the input.css
file lives.

RobinMalfait committed:
improve fast-path check

We will only call PostCSS if it contains `@import`, but now we also need
to call it when `@plugin` or `@content` exists.

RobinMalfait commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -141,10 +141,6 @@ export async function handle(args: Result<ReturnType<typeof options>>) {
-        if (glob[0] !== '/') {
-          glob = path.join(basePath, glob)
-        }
-
```
Something to figure out in the morning @philipp-spiess 

But deleting this makes the test fail again, the reason being is that each `glob` is relative to the `input.css` file (or whatever the input is). The `input.css` file lives in `basePath`. But we use `base` instead of `basePath` when calling the `scanDir` function.

The reason we use `base` (typically the `process.cwd()`) is so that auto content detection can be executed from where you execute your command.

philipp-spiess replied:
@RobinMalfait Ah right, the two different `base`es, I remember now. 😐 

I think instead of checking for `if (glob[0] !== '/') {` we should then invert it and only append the `basePath ` for relative globs (so those starting with `./` or `../`). This should work on Windows as well since we rewrite globs with posix seperators and we use the same rule for detecting relative ones in the `postcss-fix-relative-paths` plugin.

Also maybe while we're sweeing around here, what about we keep `base` for the `process.cwd()` equivalent and use `stylesheetPath` or so for the root css path?

RobinMalfait committed:
add comments for `createWatchers`

RobinMalfait committed:
use returned object from `compile` directly

RobinMalfait committed:
`contentPaths` accept a `Vec<GlobEntry>`

RobinMalfait committed:
rename `inputFilePath` to `inputBasePath`

Because it's the base path computed from the actual `inputPath`.

E.g.: `./src/styles/input.css` will result in `./src/styles`

philipp-spiess committed:
Migrate CLI test to the new integration test setup

RobinMalfait committed:
fix pnpm-lock.yaml file

We moved the `internal-postcss-fix-relative-paths` to `devDependencies`,
but the lock file wasn't up to date yet.

philipp-spiess committed:
Add test for cli watch mode

RobinMalfait committed:
remove console.log

RobinMalfait committed:
remove `--force` flag

RobinMalfait committed:
setup watchers for all directories

Including the base directory

RobinMalfait committed:
use a full `scanDir()` build in watch mode

Proper incremental builds will be a separate PR.

RobinMalfait committed:
use `watchDirectories` abstraction to calculate all directories we want
to watch

RobinMalfait committed:
improve test naming and setup

RobinMalfait committed:
add step where we change file outside of current pwd

RobinMalfait committed:
output stdout/stderr in CI

RobinMalfait committed:
collect stdout and stderr, only print output when tests fail

RobinMalfait committed:
tmp

RobinMalfait committed:
add debug logs

philipp-spiess committed:
Write a file in the right repo

RobinMalfait committed:
Revert "tmp"

This reverts commit 8558f6ea1beff90bba85a4fc787508a726557d59.

RobinMalfait committed:
Revert "add debug logs"

This reverts commit fa3f288aa84a776fceccacf63910de415c65c0c7.

RobinMalfait committed:
add `waitForOutputFileChange`

Instead of relying on CLI output, let's rely on changes made to the File
System itself.

The API looks like:

```
await waitForOutputFileChange(path, cb => {
  // Do something
})
```

It will setup a watcher for the given `path`, once the watcher is ready,
we will invoke the callback.

Once the file on disk changes, we can resolve the wrapping promise.

We also have this helper in the v3 codebase, a rename might be good,

Let's see if this fixes flakey tests on CI or not.

RobinMalfait committed:
accept return value from inner callback

RobinMalfait committed:
wait for disk file change on initial spawn

philipp-spiess committed:
Introduce fs.expectFileToContain helper
